pkg/cloud/cloud.go

```diff
@@ -408,47 +409,17 @@ func (c *cloud) DetachDisk(ctx context.Context, volumeID, nodeID string) error {
 
 // WaitForAttachmentState polls until the attachment status is the expected value.
 func (c *cloud) WaitForAttachmentState(ctx context.Context, volumeID, state string) error {
-	// Most attach/detach operations on AWS finish within 1-4 seconds.
-	// By using 1 second starting interval with a backoff of 1.8,
-	// we get [1, 1.8, 3.24, 5.832000000000001, 10.4976].
-	// In total we wait for 2601 seconds.
-	backoff := wait.Backoff{
-		Duration: 1 * time.Second,
-		Factor:   1.8,
-		Steps:    13,
-	}
-
-	verifyVolumeFunc := func() (bool, error) {
-		request := &ec2.DescribeVolumesInput{
-			VolumeIds: []*string{
-				aws.String(volumeID),
-			},
-		}
-
-		volume, err := c.getVolume(ctx, request)
-		if err != nil {
-			return false, err
-		}
-
-		if len(volume.Attachments) == 0 {
-			if state == "detached" {
-				return true, nil
-			}
-		}
-
-		for _, a := range volume.Attachments {
-			if a.State == nil {
-				klog.Warningf("Ignoring nil attachment state for volume %q: %v", volumeID, a)
-				continue
-			}
-			if *a.State == state {
-				return true, nil
-			}
-		}
-		return false, nil
+	describeVolumesInput := &ec2.DescribeVolumesInput{
+		VolumeIds: []*string{aws.String(volumeID)},
+	}
+	switch state {
+	case "attached":
+		return c.ec2.WaitUntilVolumeInUseWithContext(ctx, describeVolumesInput)
```

Comment:
By double looking at the method's implementation, it is checking for if `volumes[].State` is `InUse` state, and it is different to our current check on `Volume.Attachment[].State`. Is it okay?

Reply:
Yes this is a great point.  I think it comes down to whether the `attaching` attachment state would return as `InUse` just as `attached` does and whether that matters to us.  I'll try and test via EC2.

Reply:
I confirmed that while a volume is in the `attaching` state, that its state is reflected as `InUse` (see below).  This might be a problem for `AttachDisk` in that the native `WaitUntilVolumeInUseWithContext` function could return success before a problem happens during attachment.  I think we will want to stay with the custom check on `Volume.Attachment[].State` for the attachments.  For `CreateDisk` and `DeleteDisk` I think we can use the AWS native functions.  What do you think?

```
{
    "Volumes": [
        {
            "AvailabilityZone": "xxx",
            "Attachments": [
                {
                    "AttachTime": "2019-05-23T21:53:04.000Z",
                    "InstanceId": "xxx",
                    "VolumeId": "xxx",
                    "State": "attaching",
                    "DeleteOnTermination": true,
                    "Device": "/dev/xvda"
                }
            ],
            "Encrypted": false,
            "VolumeType": "gp2",
            "VolumeId": "xxx",
            "State": "in-use",
...
```

Reply:
sounds good to me. Let's just do this for `CreateVolume` then, we probably don't need the one for `DeleteVolume` since we are not waiting for it even now.